PATCH-P0-FEATURE-UPSTREAM: Keep datadir across multiple calls

mysql_upgrade script asks for datadir multiple times during update but at some
point privileges gets updated and if --skip-grant-tables was used (like in SUSE
init scripts), datadir is no longer queryable. So we cache the value.

Maintainer: Michal Hrusecky <Michal.Hrusecky@opensuse.org>

Index: client/upgrade/program.cc
===================================================================
--- client/upgrade/program.cc.orig
+++ client/upgrade/program.cc
@@ -170,11 +170,6 @@ public:
       }
     }
 
-    if (this->run_sql_fix_privilege_tables() != 0)
-    {
-      return EXIT_UPGRADING_QUERIES_ERROR;
-    }
-
     if (!this->m_upgrade_systables_only)
     {
       this->print_verbose_message("Checking databases.");
@@ -189,6 +184,11 @@ public:
       }
     }
 
+    if (this->run_sql_fix_privilege_tables() != 0)
+    {
+      return EXIT_UPGRADING_QUERIES_ERROR;
+    }
+
     this->print_verbose_message("Upgrade process completed successfully.");
 
     /* Create a file indicating upgrade has been performed */
@@ -319,13 +319,15 @@ private:
    */
   int get_upgrade_info_file_name(char* name)
   {
-    string datadir;
+    static string datadir;
+    if(datadir.empty()) {
     int res= Show_variable_query_extractor::get_variable_value(
       this->m_query_runner, "datadir", &datadir);
     if (res != 0)
     {
       return res;
     }
+    }
 
     fn_format(name, "mysql_upgrade_info", datadir.c_str(), "", MYF(0));
 
